src/boot/store.js

```diff
@@ -90,12 +90,7 @@ export const cacheKeys: Array<$Keys<GlobalState>> = [
 function dropCache(state: GlobalState): $Shape<GlobalState> {
   const result: $Shape<GlobalState> = {};
   storeKeys.forEach(key => {
-    // $FlowFixMe[incompatible-indexer]
-    // $FlowFixMe[incompatible-exact]
-    // $FlowFixMe[prop-missing]
-    // $FlowFixMe[incompatible-variance]
-    // $FlowFixMe[incompatible-type-arg]
-    /* $FlowFixMe[incompatible-type]
+    /* $FlowFixMe[cannot-write]
```

Comment:
> Hmm, maybe the right answer is that there should be a `ReadWrite<T>` defined in our `generics.js` that encodes that trick. Then we'd just use that here.

That sounds good.

My first thought was that maybe we could implement `ReadWrite<T>` using `$ObjMap` instead of `$Diff`:

- `$ObjMap` is singled out as having the behavior we want, in the [doc](https://flow.org/en/docs/types/utilities/#toc-readonly), while `$Diff` is not:
  > Additionally, other utility types, such as `$ObjMap<T>`, may strip any read/write annotations [...]

  (Hm, though I guess it says "may", so it's not for certain.)
- The bug in `$Diff`, which we'd be relying on for `ReadWrite`, is documented at facebook/flow#6225, and is potentially on its way to being fixed. If it got fixed, our `ReadWrite` would break.
- Someone [commented](https://github.com/facebook/flow/issues/6225#issuecomment-524046293) on that Flow issue, linking to a PR, facebook/flow#7643, which suggests that `$ObjMap` might be on its way to growing a nice feature that would help us implement `ReadWrite` quite straightforwardly, with quite a small code change.

Sadly I haven't gotten `$ObjMap` to work:

```js
type ReadWrite<T> = $ObjMap<T, <V>(V) => V>;
```

When I try to use this, Flow still complains with the `cannot-write` code; you can see this happening on the latest Flow version (alongside an implementation with `$Diff` that does work) in [this repro](https://flow.org/try/#0C4TwDgpgBAShCGATA6gJwJbAgRgDwBUA+KAXigBIB5AIwCsBZeMAgGilwDVCAKDgSlLEuAbgBQoSLAQoMWAEwFiZcgBF0AM3WsoAbwA+egL6Ex48NHzmAgqQpwklAHYAbELn1R1Aey8AuKI4ArgC21BCoUEYmoqIAxl6OAM7AUPD+9jKYOATWSrqePv7YcgDMUIZi8Ukp1OnSaFkKlpBWeToFflDFZRUx8AB03l62JQAsAKzCUAD001JIDVjYAV4pAO5eqADW6I4A5lC+3KLUgz4jE1Oz85nyUBvbuwe+fEA) (let me know if the link doesn't work).

Reply:
So I think I'll go with `$Diff` for now.

Reply:
> The `$Rest<Foo, { ... }>` trick that comes up a lot in `exampleData`, to make all the properties optional, is another one that'd be good to give that treatment to. (Came to mind because there were a bunch of instances of it in the first commit of this PR.) Maybe named `Shape`? `GoodShape`? It's basically a fixed version of `$Shape`, which is unsound (https://flow.org/en/docs/types/utilities/#toc-shape .)

Hmm, yeah, good idea. I see from f7354c1ff that `$ReadOnly<$Shape<…>>` is reportedly sound, while just `$Shape` is not. Do you think that's right, and do you think that may sensibly remove the need for a `Shape`, `GoodShape`, etc.?

Reply:
> I see from [f7354c1](https://github.com/zulip/zulip-mobile/commit/f7354c1ff3504d1ebefc5d0959d7f972ccb48c00) that `$ReadOnly<$Shape<…>>` is reportedly sound, while just `$Shape` is not.

Hmm, it does say that, but it's not clear what source or reasoning that's based on. In particular that's not mentioned in the upstream docs that [point out `$Shape` is unsound](https://flow.org/en/docs/types/utilities/#toc-shape).

Just from a quick look at the example from there in the docs and playing with that in the Flow playground, I think in fact it's not true. Here's the example, with `$ReadOnly` added:
```js
type Person = $ReadOnly<{
  age: number,
  name: string,
}>;
type PersonDetails = $ReadOnly<$Shape<Person>>;

const personShape: PersonDetails = {age: 28};
(personShape: Person);
```

That actually does give an error… but only at the `{age: 28}` literal. Which seems like its own bug -- that line ought to be fine.

And in fact if you adapt the example so it doesn't trip that bug, Flow unsoundly accepts it:
```js
(personShape: PersonDetails): Person => personShape;
```

([Playground link here.](https://flow.org/try/#0C4TwDgpgBAChBOBnA9gOygXigEgEoQEMATAeVQBsQAeAbwCgooCBzCALilQFcBbAIwQAaBpwI92URMHgBLVM2EBfAHwBuOqEiwEKVABEIwAjPKJMOfMTKUq2AMoALApCpwkaZWrp0AxmilQkO6ojs4SbroGRiZmWDQsEgBMAByK6gAUQbqhkBwRaACU6nSZOmg54WX6hsamBXlVmMqBVRXqQA))

So while we do want most of our object types to be `$ReadOnly`, and that might alleviate some of the symptoms of the unsoundness of `$Shape`, I don't think it really solves it.